File: 001-wc/wc.pl
#!/usr/bin/perl

use v5.28;
use warnings;
use warnings qw(FATAL utf8);
use autodie qw(open close);
use open qw(:encoding(UTF-8) :std);

## File name
my $file = '';

## Set based on flags passed in via command line
my ($doBytes, $doLines, $doWords, $doChars) = (0, 0, 0, 0);

## Not sure why wc outputs different spacing depending on if the file is passed or piped in
my $sizeMod = 1;

## Do we have command-line parameters?
if (scalar(@ARGV) > 0) {
    my ($arg1, $arg2) = @ARGV;
    $sizeMod = 0;

    ## Check if first option starts with '-'
    if (index($arg1, '-') == 0) {
        if ($arg1 eq '-c') {
            $doBytes = 1;
        }
        elsif ($arg1 eq '-l') {
            $doLines = 1;
        }
        elsif ($arg1 eq '-w') {
            $doWords = 1;
        }
        elsif ($arg1 eq '-m') {
            $doChars = 1;
        }
        else {
            die "Invalid command line switch: $arg1\n";
        }

        ## If we also received a file path, save it
        if (defined($arg2)) {
            $file = $arg2;
        }

        ## Remove first item from @ARGV leaving either nothing or a file path
        ## Needed so <> can do its job below to read the input
        shift;
    }
    else {
        ## No flags passed so default to -c, -l and -w
        $file = $arg1;
        $doBytes = 1;
        $doLines = 1;
        $doWords = 1;
    }
}
else {
    ## Only input from STDIN with no flags so default to -c, -l and -w
    $doBytes = 1;
    $doLines = 1;
    $doWords = 1;
}

my ($bytes, $lines, $words, $chars) = (0, 0, 0, 0);

## Read the file we left in @ARGV or from STDIN
while (<>) {
    $lines++;
    $chars += length;
    $bytes += bytes::length;
    $words++ while /\S+/g;
}

my $bytesSize = length($bytes);
my $output = '';

if ($doLines) {
    $output .= sprintf("%*d ", ($doBytes ? $bytesSize : length($lines)) + $sizeMod, $lines);
}
if ($doWords) {
    $output .= sprintf("%*d ", ($doBytes ? $bytesSize : length($words)) + $sizeMod, $words);
}
if ($doBytes) {
    $output .= sprintf("%*d ", ($bytesSize + $sizeMod), $bytes);
}
if ($doChars) {
    $output .= sprintf("%d ", $chars);
}

$output .= $file;

print "$output\n";

File: 001-wc/wc.lua
#!/usr/bin/lua

local arg1, arg2 = ...
local file = ''
local doBytes, doLines, doWords, doChars = 0, 0, 0, 0
local sizeMod = 1;

-- Only input from STDIN with no flags so default to -c, -l and -w
if arg1 == nil then
    doBytes, doLines, doWords = 1, 1, 1
else
    -- Check if first option starts with '-'
    if string.sub(arg1, 1, 1) == '-' then
        if arg1 == '-c' then
            doBytes = 1
        elseif arg1 == '-l' then
            doLines = 1
        elseif arg1 == '-w' then
            doWords = 1
        elseif arg1 == '-m' then
            doChars = 1
        else
            print(string.format("Invalid command line switch: %s", arg1))
            os.exit(1)
        end

        if arg2 ~= nil then
            file = arg2
        end
    else
        -- No flags passed so default to -c, -l and -w
        file = arg1
        doBytes, doLines, doWords = 1, 1, 1
    end

    sizeMod = 0
end

local fh = (file == '' and io.stdin) or io.open(file, 'rb')
local bytes, lines, words, chars = 0, 0, 0, 0

for line in fh:lines('L') do
    bytes = bytes + line:len()
    lines = lines + 1
    _, n = line:gsub("%S+", "")
    words = words + n
    chars = chars + utf8.len(line)
end

fh:close()

local bytesSize = string.len(bytes)
local output = ''
local fmt = ''

if doLines == 1 then
    fmt = ("%s%_d "):gsub("_", (doBytes == 1 and bytesSize or string.len(lines)) + sizeMod)
    output = string.format(fmt, output, lines)
end

if doWords == 1 then
    fmt = ("%s%_d "):gsub("_", (doBytes == 1 and bytesSize or string.len(words)) + sizeMod)
    output = string.format(fmt, output, words)
end

if doBytes == 1 then
    fmt = ("%s%_d "):gsub("_", bytesSize + sizeMod)
    output = string.format(fmt, output, bytes)
end

if doChars == 1 then
    output = string.format("%s%d ", output, chars)
end

output = output .. file

print(output)

